valhalla/meili/routing.h

```diff
@@ -148,62 +179,102 @@ class LabelSet
  public:
   LabelSet(const float max_cost, const float bucket_size = 1.0f);
 
-  bool put(const baldr::GraphId& nodeid, sif::TravelMode travelmode,
-           std::shared_ptr<const sif::EdgeLabel> edgelabel) {
-    return put(nodeid, {},                 // nodeid, (dummy) edgeid
-               0.f, 0.f,                   // source, target
-               sif::Cost(), 0.f, 0.f,      // cost, turn cost, sort cost
-               baldr::kInvalidLabel,       // predecessor
-               nullptr, travelmode, edgelabel);
+  /**
+   * Add an origin label using a destination index.
+   */
+  void put(const uint16_t dest, const sif::TravelMode mode,
+           const Label* edgelabel) {
+    // Do not add a duplicate label for the same destination index
+    if (dest_status_.find(dest) == dest_status_.end()) {
+      // If edgelabel is not null, append it to the label set otherwise append
+      // a dummy. In both cases add the label to the priority queue, set its
+      // predecessor to kInvalidLabel, and initialize costs to 0.
+      const uint32_t idx = labels_.size();
+      queue_->add(idx, 0.0f);
+      dest_status_.emplace(dest, idx);
+      labels_.emplace_back(edgelabel ? *edgelabel : Label());
+      Label& label = labels_.back();
+      label.InitAsOrigin(mode);
+      label.set_dest(dest);
+      label.set_nodeid({});
+    }
   }
 
-
-  bool put(const baldr::GraphId& nodeid,
-           const baldr::GraphId& edgeid,
-           float source, float target,
-           const sif::Cost& cost, float turn_cost, float sortcost,
-           uint32_t predecessor,
-           const baldr::DirectedEdge* edge,
-           sif::TravelMode travelmode,
-           std::shared_ptr<const sif::EdgeLabel> edgelabel);
-
-  bool put(uint16_t dest, sif::TravelMode travelmode,
-           std::shared_ptr<const sif::EdgeLabel> edgelabel) {
-    return put(dest, {},                  // dest, (dummy) edgeid
-               0.f, 0.f,                  // source, target
-               sif::Cost(), 0.f, 0.f,     // cost, turn cost, sort cost
-               baldr::kInvalidLabel,      // predecessor
-               nullptr, travelmode, edgelabel);
+  /**
+   * Add an origin label using a node id.
+   */
+  void put(const baldr::GraphId& nodeid, const sif::TravelMode mode,
+           const Label* edgelabel) {
+    // Do not add a duplicate origin label for the same node
+    if (node_status_.find(nodeid) == node_status_.end()) {
```

Comment:
It seems that we have the same logic twice. Can we reuse the `void put(const baldr::GraphId& nodeid, ...)` below? Most params can be obtained from `edgelabel`.